rfc/api.md

```diff
@@ -0,0 +1,120 @@
+
+This complex graph structure requires more runtime checks which can be theoretically replaced with static checks, which gives us the following API.
+
+Transmission channel does step forward after invoking `send` method, which accepts event type and its arguments described in protocol `dispatch_type`. It returns the following transmission state, which can be reassigned to the old `tx`, because it's useless now, which gives us an ability to chain requests:
+
+After termination leaf reached you is unable send more chunks, because there won't be such method in terminated sender.
+
+```
+tx = tx.send<C>(args...).get(); // Simple send, can throw if unable to properly pack given arguments or any network error received.
+
+tx.send<C1>(args1...).get().send<C2>(args2...).get().send<C3>(args3...).get(); // Chained.
+```
+
+In the opposite side there is a received, from which you can receive and extract incoming chunks. Unfortunately it should return both the new receiver type and chunk requested.
+
+```
+std::tie(rx, result) = rx.recv().get(); // Can throw on any network error.
+```
+
+In the common case result type is an algebraic data type, which consists of types described in Cocaine Service's protocol `upstream_type`.
+
+At this moment there are only two possible result types: option and stream. They can be described as:
+
+```
+option<T> ::= T | tuple<int, string>.
+stream<T> ::= T | tuple<int, string> | choke.
+```
+
+All described above means that after receiving a result from the receiver future all you need is to visit obtained variant type.
+
+I've just described the commonest protocol description I can imagine. One have imagined it's extremely complex API.
+
+The good news is it can be simplified for 95% cases.
+
+If you work with primitive `upstream_type`and with void `dispatch_type` which the most services are, then the calling interface can be simplified to:
+
+```
+// Returns a value on success, throws `cocaine_error` on service error or `system_error` on network error.
+try {
+    service.invoke<Event>(args...).get(); 
+} catch (const cocaine_error& err) {
+    // ...
+} catch (const std::system_error& err) {
+    // ...
+}
+```
+
+If you work with streaming `upstream_type` and with `streaming<std::string>` `dispatch_type` which all the user applications are, then the interface could be represented as:
+
+```
+std::tie(tx, rx) = service.invoke<cocaine::io::app::enqueue>("event");
+tx.send<protocol::chunk>("le message").get()
+  .send<protocol::error>(-1, "the user gone mad").get();
+
+auto result = rx.recv();    // -> variant<T, E, C> where T == dispatch_type, E - error type, C - choke.
+auto result = rx.recv<T>(); // -> optional<T> or throws `cocaine_error` on error.
```

Comment:
Yes, `rx` isn't invalidated after `recv()` for streaming services, which all applications are. In that case you cannot predict, when the app returns ERROR/CHOKE, which means stream termination and `rx` invalidation of course. 

In the case of `tx`, you **can**, because you own it, and I can guarantee its type safety. That's why `tx` doesn't work the same way.